### PR TITLE
Use global workers to reduce goroutine usage for gateway stats.

### PR DESCRIPTION
**Pull request by guoliang100**
@alainjobart @sougou 

Benchmark shows the latency to (non-blocking) send into a buffered chan is relatively constant regardless of the number of senders.

---
This change is [Reviewable](https://reviewable.io/reviews/youtube/vitess/1604)

### CODE REVIEW

File: go/vt/vtgate/gateway.go [hunk 1]
```diff
@@ -128,21 +128,55 @@ type GatewayEndPointCacheStatus struct {
...
+func processQueryInfo() {
+	for qi := range aggrChan {
+		qi.aggr.processQueryInfo(qi)
+	}
+}
+
+// NewGatewayEndPointStatusAggregator creates a GatewayEndPointStatusAggregator.
+func NewGatewayEndPointStatusAggregator() *GatewayEndPointStatusAggregator {
+	gepsa := &GatewayEndPointStatusAggregator{}
 	go func() {
```
Comment (alainjobart) [inline go/vt/vtgate/gateway.go]: I think this one can be in the function, why use a go routine here?

File: go/vt/vtgate/gateway.go [hunk 2]
```diff
@@ -128,21 +128,55 @@ type GatewayEndPointCacheStatus struct {
...
+	aggrChan chan *queryInfo
+	// muAggr protects below vars.
+	muAggr sync.Mutex
+	// aggregators holds all Aggregators created.
+	aggregators []*GatewayEndPointStatusAggregator
+)
+
+func init() {
+	// init global goroutines to aggregate stats.
+	aggrChan = make(chan *queryInfo, 100000)
```
Comment (alainjobart) [inline go/vt/vtgate/gateway.go]: let's make the 100k and the following 10 below constants at the beginning of the file.

**Comment by alainjobart** [comment]
The whole point of all this code is to asynchronously aggregate the queryInfo objects, right?

It seems to me we should first try to just do it synchronously. Do we have benchmarks that show that doing all this asynchronously is better? Lock contention on each aggregator is probably *very* small, as a single vtgate handles 1k QPS or so, to 500 tablets or so, I am pretty sure the stats per endpoint are not contributing too much to tail latency?

(the work being done by each processQueryInfo is fairly small: couple increments, couple very small map access, ...).

Let's try to simplify first, then if we get proof this is an issue, we can make it more complex?

**Comment by alainjobart** [comment]
Just checked a production vtgate, QPS is actually 2 or so on each aggregator. So my numbers were correct. lock contention on these guys will be so small processing the stats synchronously is the way to go.

**Comment by guoliang100** [comment]
We benchmarked the sync version before. When qps is high, it shows a 1ms difference on 99th. That is why I changed it to async.

**Comment by alainjobart** [comment]
I guess we need to support a vtgate with a small number of high QPS endpoints, so even if it's not our production setup, it's worth it overall. I'm a bit worried we don't understand all this very well... let's talk about this in person.

Looking at this code some more, I think we can do a bit better:
- we use 60 buckets of 1 second. That makes the rotation of buckets very frequent. How about using 6 buckets of 10s instead?
- your code uses 10 channel reader go routines, could we start with just one? I don't think we'll need more than one core to process the stats?
- small nit in processQueryInfo (independent of this CL): TabletType should only be set if it's different in the previous if clause.

File: go/vt/vtgate/gateway.go [hunk 3]
```diff
@@ -176,19 +210,21 @@ type queryInfo struct {
...
+		aggr:       gepsa,
 		addr:       addr,
 		tabletType: tabletType,
 		elapsed:    elapsed,
 		hasError:   hasError,
 	}
 	select {
-	case gepsa.qiChan <- qi:
+	case aggrChan <- qi:
 	default:
```
Comment (alainjobart) [inline go/vt/vtgate/gateway.go]: let's count the cases where we drop one of these with an atomic int.

**Comment by alainjobart** [comment]
OK after talking about it in person (thanks!) we agreed to move forward with this. Switching back to a single Mutex is not good, as it would add tail latency in the case the mutex is busy, and go scheduler un-schedules a waiting go routine (and doesn't re-schedule it for a while). We explored using a RWMutex, and only taking the Write lock when we change the buckets. This is complicated as well, and we rotate frequently enough that it might be an issue.

Changes to make:
- add an atomic int to count & report the number of times we drop a stat because the channel is full.
- this will allow us to lower the 100k buffer size to a smaller number as we'll know when it's too small.
- also, reduce the number of consumer go routines to one. Again the atomic int will help track when we don't keep up, if it ever happens.

File: go/vt/vtgate/gateway.go (hunk 1 above)
Comment (guoliang100) [inline go/vt/vtgate/gateway.go]: Removed the goroutine.

File: go/vt/vtgate/gateway.go (hunk 2 above)
Comment (guoliang100) [inline go/vt/vtgate/gateway.go]: Done

File: go/vt/vtgate/gateway.go (hunk 3 above)
Comment (guoliang100) [inline go/vt/vtgate/gateway.go]: Done

**Comment by guoliang100** [comment]
@alainjobart 

Updated the code as discussed. Please take a look.

**Comment by alainjobart** [comment]
LGTM

**Pull request merged**